106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one and also adds the `spec.mode` to it.
+The generated `KafkaRebalance` custom resource will be called `<my-cluster-name>-auto-rebalancing-skew-anomalyId`, where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, and `skew` refers to applying the rebalance to all the brokers and the anomalyId would be retrieved from the notifier.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-cluster-auto-rebalancing-add-brokers
+  finalizers:
+    - strimzi.io/auto-rebalancing
+spec:
+  mode: skew
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The operator also sets a finalizer, named `strimzi.io/auto-rebalancing`, on the "actual" `KafkaRebalance` custom resource.
+This is needed to avoid the user, or any other tooling, to delete the resource while the auto-rebalancing is still running.
+The finalizer is removed at the end of the auto-rebalancing process, with or without errors, allowing the "actual" `KafkaRebalance` custom resource deletion by the operator itself.
+
+#### AnomalyDetectorNotifier
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()` etc.
+The `AnomalyDetectorNotifier` will be based of the `AnomalyNotifier` class.
+The anomalies with smaller priority value and detected time will be considered priority and resolved first.
+In case the anomaly is unfixable due to issues like  lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy ReplicaCapacityGoal, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored.
+The `AnomalyDetectorNotifier` will override all the abstract methods provided by the `AnomalyNotifier` interface.
+We will add an `alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create a configmap with name set as `goal-violation` followed by `anomalyId`.
+
+The ConfigMap would look like this:
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: goal-violation-<anomalyID>
+data:
+  anomalyType: GOAL_VIOLATION
+  anomalyId: "<anomaly-id>"
+  detectionTime: "<time-of-detection>"
+# ...
+```
+
+The operator will then check if any configmap with prefix `goal-violation` is created or not, if it finds one created then operator will trigger the rebalance.
```

Comment:
```suggestion
The operator will then check if any ConfigMap with prefix `goal-violation` is created or not, if it finds one created then operator will trigger the rebalance.
```

Reply:
I know it is not recommended but what if a user deploys multiple Kafka clusters in the same namespace? How does the operator know what kafka cluster this CM relates too?